test/java/integration/AndroidEnvoyFlowTest.java

```diff
@@ -0,0 +1,356 @@
+package integration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import android.content.Context;
+import androidx.test.core.app.ApplicationProvider;
+import io.envoyproxy.envoymobile.AndroidEngineBuilder;
+import io.envoyproxy.envoymobile.Engine;
+import io.envoyproxy.envoymobile.EnvoyError;
+import io.envoyproxy.envoymobile.RequestHeaders;
+import io.envoyproxy.envoymobile.RequestHeadersBuilder;
+import io.envoyproxy.envoymobile.RequestMethod;
+import io.envoyproxy.envoymobile.ResponseHeaders;
+import io.envoyproxy.envoymobile.ResponseTrailers;
+import io.envoyproxy.envoymobile.Stream;
+import io.envoyproxy.envoymobile.UpstreamHttpProtocol;
+import io.envoyproxy.envoymobile.engine.AndroidJniLibrary;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import okhttp3.mockwebserver.Dispatcher;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
+public class AndroidEnvoyFlowTest {
+
+  private static Engine engine;
+
+  private final MockWebServer mockWebServer = new MockWebServer();
+
+  @BeforeClass
+  public static void loadJniLibrary() {
+    AndroidJniLibrary.loadTestLibrary();
+  }
+
+  @AfterClass
+  public static void shutdownEngine() {
+    if (engine != null) {
+      engine.terminate();
+    }
```

Comment:
Do we need to reset the engine to null here since we are doing a null check for initialization in the `@Before` step?

Reply:
Well, for sure, the tests are fast: no start & stop after each @Test. 

I assumed that doing a new `AndroidEngineBuilder(appContext)....build();` for each test would not work if we call `engine.terminate();` after each @Test.